src/aoc2020/day3_update.clj

```diff
@@ -0,0 +1,103 @@
+(ns aoc2020.day3-update
+  (:require [aoc2020.util :refer [file->seq]]))
+
+(defn grid-template->grid
+  [grid-template]
+  (mapv (fn [row] (cycle (seq row)))
+        grid-template))
+
+(defn position-s
+  [{slope-right :right
+    slope-down  :down}]
+  (pmap (fn [x]
+          [(* slope-down x) (* slope-right x)])
+        (range)))
+
+(defn value-at-pos
+  [grid [row col]]
+  (let [row-s (take (+ 1 col) (grid row))]
+    (nth row-s col)))
+
+(defn tree-count
+  [grid-template slope-dist]
+  (let [row-number (count grid-template)
+        grid       (grid-template->grid grid-template)]
+    (->> slope-dist
+         (position-s)
+         (take-while (fn [pos] (< (first pos) row-number)))
+         (map (partial value-at-pos grid))
+         (filter (fn [value] (= \# value)))
+         (count))))
+
+
+
+(defn product-of-tree-count
+  [grid-template slope-dist-s]
+  (->> (map (fn [slope-dist]
+              (tree-count grid-template slope-dist))
+            slope-dist-s)
+       (reduce *)))
+
+
+(comment
+  (do (def sample-grid-template (file->seq "aoc2020/day3/sample-input.txt")) sample-grid-template)
+  #_=> ["..##......."
+        "#...#...#.."
+        ".#....#..#."
+        "..#.#...#.#"
+        ".#...##..#."
+        "..#.##....."
+        ".#.#.#....#"
+        ".#........#"
+        "#.##...#..."
+        "#...##....#"
+        ".#..#...#.#"]
+
+  (do (def grid-template (file->seq "aoc2020/day3/input.txt")) grid-template)
+  #_=> ["...........#..#.#.###....#....."
+        "...#..#...........#.#...#......"
+        "#.....#..#........#...#..##...."
+        "..#...##.#.....#.............#."
+        "#.#..#......#.....#....#......."
+        ".....#......#..#....#.....#...."
+        ".......#.#..............#......"
+        ".....#...#..........##...#.....",,,]
+
+  ;; part 1
+
+  (def slope-dist {:right 3 :down 1})
+
+  ;; generate infinite grid
+  (def sample-grid (grid-template->grid sample-grid-template))
+
+  ;; generate the position[row col] infinite sequence
+  (position-s slope-dist)
+  #_=> ([0 0] [1 3] [2 6] [3 9] [4 12] ...)
+
+  ;; get the value of the position[row col]
+  (value-at-pos sample-grid [0 0])
+  #_=> \.
+
+  (value-at-pos sample-grid [1 0])
+  #_=> \#
+
```

Comment:
Where is the sequence of values corresponding to the positions?

Reply:
The sequence is infinite, I need to `take` a part of that and then I can map `value-at-pos` to sequence. I will add that.